BookStore Manager and MarlinBookConfig added

JBenda opened the pull request:
BEGINRELEASENOTES

### MarlinBook
+ fixed spelling, add documentation
+ added `WeakEntry BookStore::findFirst(Condition)` for single element access
+ added Masks to Flags
+ added bitwise not operator `~` to Flags
+ if concurrency = 1 book everything with the Single Memory layout
+ throw an exception if Single memory layout is used with concurrency != 1
+ no output file if there is no object to store

### Framework
+ Added MarlinBookConfig, includes, and typedefs for MarlinBook
+ Added BookStoreManager, 
   - handle BookStore for Application
   - write to disk with `BookStoreManager::writeToDisk()`   
   -  provides Interface
      * bookHist[Type][Dimenison]
      * getHist[Type][Dimension]
+ moved concurrency calculation in Application
+ added parameter node `<bookstore>` for BookStore parameters
+ added bookstore parameter "OutputFile" to write BookStore results in this File.
   if not set generates a file with foramt MarlinMT_[PId].root
+ added bookstore parameter "DefaultMemeoryLayout" to set default memory layout
   values: Copy, Share, Default (aka Share)
+ booked objects will be stored if not told different
+ global parameter "Concurrency" now defaults to "auto"


### Test enviroment
+ activated MarlinBook test in travis-ci


ENDRELEASENOTES

rete committed:
Added generated Doxyfile to gitignore

JBenda committed:
Add a variety  of store to BookStore

+ storeList: stores each Entry which key is listed
+ storeSelection: stores custom created selection

also:
+ fixed spelling
+ rename EntryKey.hash to EntryKey.idx (because this is what it is)

rete committed:
Added new typedefs

rete committed:
Corrected usage of book namespace

rete committed:
Fixed AxisConfig typedef usage

rete committed:
Added new class BookStoreManager

JBenda committed:
implimented first histogram function

JBenda committed:
Implemented first histogram function

JBenda committed:
Changed concurrency default to auto

JBenda committed:
first running processor with save function

JBenda committed:
Make Book Manager interface more generic

JBenda committed:
Wrap instantiation code in Macro

JBenda committed:
Implemented Macro and fix minor errors

JBenda committed:
Added more Histogram constructions

JBenda commented on cmake/MarlinBookConfig.h.cmake.in:
```diff
@@ -1,62 +1,69 @@
-
+#pragma once
```
for not double include configs and header from MarliBook

JBenda commented on cmake/MarlinBookConfig.h.cmake.in:
```diff
@@ -1,62 +1,69 @@
+#include <marlin/book/Hist.h>
 
 namespace marlin {
-  
```
just change indent

JBenda commented on source/book/include/marlin/book/BookStore.h:
```diff
@@ -75,6 +75,10 @@ namespace marlin {
 
       const T& merged() const ;
 
+      [[nodiscard]] const EntryKey& key() const {
```
new getter for `Handle<Entry<T>>` to access key 
will be used in BookStoreManager for storing objects (in files)

JBenda commented on source/book/include/marlin/book/BookStore.h:
```diff
@@ -83,7 +87,7 @@ namespace marlin {
       std::atomic< std::size_t > _count{0} ;
       /// shared mutex for serialisation when writing 
-      std::shared_mutex _mappingAcces;
+      std::shared_mutex _mappingAcces{};
```
Added default constructor (fixed Warning)

rete replied:
`_mappingAcces` -> `_mappingAccess`

JBenda commented on source/book/include/marlin/book/BookStore.h:
```diff
@@ -126,20 +130,33 @@ namespace marlin {
        *  @throw BookStoreException key not exist in Store.
        */
-      details::Entry &get( const EntryKey &key ) { return get( key.hash ); }
+      details::Entry &get( const EntryKey &key ) const { return get( key.idx ); }
```
fixed added const modifier,
rename key.hash to idx, because it is an id and not an hash

rete replied:
If this is const why don't you return a const reference to the `details::Entry` ?

JBenda replied:
is for internal usage (private), because the BookStore only handles pointer to Entry it is no problem, I use it in the `remove(key)` function, like: `get(key).delete()`, which is with const Entry not possible

rete replied:
If it returns a reference without const, you should remove the const from the function signature

JBenda commented on source/book/include/marlin/book/BookStore.h:
```diff
@@ -126,20 +130,33 @@ namespace marlin {
+      details::Entry &get( std::size_t const idx ) const {
         try {
-          return *_entries[key] ;
+          return *_entries[idx] ;
```
renaming hash to idx, see above for details

JBenda commented on source/book/include/marlin/book/BookStore.h:
```diff
@@ -126,20 +130,33 @@ namespace marlin {
+          MARLIN_BOOK_THROW( "Invalid key." ) ;
+        }
+      }
+
```
make EntryPointer accessible for cleaner usage in `Handle<Entry<T>>` construction

rete replied:
Documentation string ?

JBenda replied:
What do you mean by this?
the throw is documented with \@throw above for Doxygen ...

JBenda commented on source/book/include/marlin/book/BookStore.h:
```diff
@@ -238,11 +255,23 @@ namespace marlin {
       /**
        *  @brief saves all Objects in one Root-File. 
-       *  @param path where save Root-File. 
+       *  @param writer used to serialize objects. 
```
update function discription

JBenda commented on source/book/include/marlin/book/BookStore.h:
```diff
@@ -238,11 +255,23 @@ namespace marlin {
+      template<typename Itr>
+      void storeList( StoreWriter& writer, Itr begin, Itr end) const ; 
+
+      void storeSelection( StoreWriter& writer, const Selection& selection ) const ;
```
new store functions for more flexbility

rete replied:
Documentation string ?

JBenda commented on source/book/include/marlin/book/EntryData.h:
```diff
@@ -35,7 +35,7 @@ namespace marlin {
       Flag_t flags{} ;
       /// unique number for Entry
-      std::size_t hash{0} ;
+      std::size_t idx{std::numeric_limits<std::size_t>::max()} ;
```
renaming hash -> idx

JBenda commented on source/book/include/marlin/book/MemLayout.h:
```diff
@@ -89,16 +89,16 @@ namespace marlin {
       SharedMemLayout( SharedMemLayout && ) noexcept            = default ;
       SharedMemLayout &operator=( SharedMemLayout && ) noexcept = default ;
-      ~SharedMemLayout() final                                  = default ;
+      ~SharedMemLayout() override final                         = default ;
```
final don't implicitly override, tow keywords for tow things to notice.

final just says that this function can't be overriden, but don't need to overwite a function.

rete replied:
I'm surprised this was not failing with `Werror` before. We've turned ON this warning (missing override)...

JBenda replied:
because in travis we compiled until now with MARLIN_BOOK_IMPL=dummy -> this code won't be compiled. I changed this now.

JBenda commented on source/book/include/marlin/book/configs/Base.h:
```diff
@@ -309,6 +309,19 @@ namespace marlin {
+        throw "numeber cast with data lose!";
+      }
+      
+    } // end namespace details
```
save casting of numbers to smaller spaces, 
ROOT use int for size types which don't fit with std who used size_t (unsigned long)

JBenda commented on source/framework/include/marlin/Application.h:
```diff
@@ -217,8 +233,13 @@ namespace marlin {
     ///< Whether the currently pushed event is the first one
     bool                       _isFirstEvent {true} ;
+    ///< number of concurrency with which the application runs.
+    std::size_t                _concurrency {0} ;
```
saves concurrency number, to be accessed from every Manager

JBenda commented on source/framework/include/marlin/BookStoreManager.h:
```diff
@@ -0,0 +1,85 @@
+        const Processor *proc,
+        const std::filesystem::path &path ) ;
+
+    template<typename T>
```
get Object function for internal usage

rete replied:
I usually optional for user return types to let them handle an error. Internally I would use exceptions rather than optional. Any thought on this ?

JBenda replied:
throw is way better, because of optimisation, I will change it.

User should also get an error, because it's faster and easyer to handle.

JBenda commented on source/framework/src/Application.cc:
```diff
@@ -388,4 +391,50 @@ namespace marlin {
+
+  //--------------------------------------------------------------------------
+  
+  std::size_t Application::readConcurrency() {
```
moved concurrency calculation to Application

JBenda commented on source/framework/src/Processor.cc:
```diff
@@ -13,9 +13,9 @@ namespace marlin {
     _logLevelName("") {
     // verbosity default parameter
     registerOptionalParameter( "Verbosity" ,
-			       "verbosity level of this processor (\"DEBUG0-4,MESSAGE0-4,WARNING0-4,ERROR0-4,SILENT\")"  ,
```
fixed indents

JBenda commented on source/framework/src/XMLParser.cc:
```diff
@@ -328,43 +329,43 @@ namespace marlin{
         //
         //     else if( procCount > typedProcCount ){
         //       std::cout  << "---------------------------------------------------------------------" << std::endl
-        // 		 << "  WARNING XMLParser : some of the available processors don't have    " << std::endl
```
sorry about this,.. but ther where tabs.

JBenda commented on source/framework/src/concurrency/PEPScheduler.cc:
```diff
@@ -126,30 +126,9 @@ namespace marlin {
-      if ( ccy == 1 ) {
-        _logger->log<WARNING>() << "-- The program will run on a single thread --" << std::endl ;
-        // TODO should we throw here ??
-      }
```
moved to Appliction

JBenda commented on test/tests/test-book-store-store.cc:
```diff
@@ -243,32 +243,58 @@ int main( int /*argc*/, char * /*argv*/[] ) {
+    } catch ( const marlin::book::exceptions::BookStoreException &expc ) {
+      test.test( std::string( "unexpected error: '" ) + expc.what() + "'",
+                 false ) ;
+    }
```
minor changes to adapte to new functions.

JBenda committed:
Added BookStore test to travis

+ builds now with MARLIN_BOOK_IMPL=root7

JBenda committed:
removed unfinished test

rete commented on source/book/include/marlin/book/configs/Base.h:
```diff
@@ -309,6 +309,19 @@ namespace marlin {
+            && static_cast<O>(input) <= std::numeric_limits<O>::max()) {
+          return static_cast<O>(input);
+        }
+        throw "numeber cast with data lose!";
```
typo + use exception class

JBenda replied:
fixed :)

rete commented on source/framework/include/marlin/Application.h:
```diff
@@ -142,6 +143,11 @@ namespace marlin {
+    /**
+     *  @brief Get book store manager
+     */
+    BookStoreManager &bookStoreManager() const ;
```
const returns a const reference, else no const returns a reference. Don't mix the two, make it clear.

JBenda replied:
changed return type to  `const std::unique_ptr<BookStoreManager>&` because the Applications owns the BookStoreManager, but the Manager can change, also when the Application is const.

rete commented on source/framework/include/marlin/BookStoreManager.h:
```diff
@@ -0,0 +1,85 @@
+    using Logger = Logging::Logger ;
+    
+  public:
+    BookStoreManager() { std::cout << "BSM constructed:!";} ;
```
Never use cout. the global macro for logging stuff is `streamlog_out(LEVEL)` instead of cout, i.e  `streamlog_out(MESSAGE)`. It requiries to include Logging.h also

JBenda replied:
I will remember, this should be deleted anyway.

rete commented on source/framework/src/BookStoreManager.cc:
```diff
@@ -0,0 +1,102 @@
+    book::Selection res = _bookStore.find(
+      book::ConditionBuilder()
+        .setType(typeid(T))
+        .setPath(constructPath(proc, path).string()) // TODO: life time check
```
Yes please ! :-)

JBenda committed:
fix spelling, add documentation

rete commented on source/book/include/marlin/book/BookStore.h:
```diff
@@ -123,19 +127,14 @@ namespace marlin {
       /**
-       *  @brief get Entry from key.
-       *  @throw BookStoreException key not exist in Store.
+       *  @brief 
```
Doxygen is incomplete

JBenda replied:
added param and return now

rete commented on source/framework/include/marlin/ProcessorApi.h:
```diff
@@ -56,45 +51,203 @@ namespace marlin {
+        const AxisConfigD &axisconfigX,
+        const AxisConfigD &axisconfigY,
+        const BookFlag &flags  = DefaultConfiguration) ; 
+
```
Is `bookHist3F()` missing on purpose ?

JBenda replied:
no, sorry, just miss it, will add H1I, H2I and H3I also

rete commented on source/framework/src/BookStoreManager.cc:
```diff
@@ -0,0 +1,159 @@
+  //--------------------------------------------------------------------------
+    
+  
+  BookStoreManager::~BookStoreManager() {
```
Bad idea. Not good to write the book store content on destruction. Create a function to do it. Users should not be forced to write data to disk

JBenda replied:
he isn't because it's only writing what is set to write.

But yes an extra function is more elegant :)

rete replied:
>he isn't because it's only writing what is set to write.

Yes but we don't want a side effect of writing data to disk without asking explicitly for it.
Create a function but do not call it in the destructor. It should be empty for now

JBenda committed:
+F Selection find first; use throw instead of opt

JBenda committed:
Move write to disk functionality

no `BookStoreManager::writeToDisk()` explicit called by `Application`

JBenda committed:
Added global Parameter for BookStore output file

`StoreOutputFile`

JBenda committed:
BookStoreManager: Added book and get functions

**now available** float|double|int 1|2|3

JBenda committed:
Added Doxygen description for BookStore::getPtr

JBenda committed:
changed function signature, more clear now

`Application::bookStoreManager` changed return type from non const ref
to const ref unique pointer.

JBenda committed:
changed safe_cast exception

* now throws a std::domain_error
* static assert for no integral types

JBenda committed:
Fixed const correctness problems

* `Entry &BookStore::get(…)const` -> `const Entry&Bookstore::get(…)const`
* added `Entry& BookStore::get(…)`
* `BookStoreManager& Application::bookStoreManager() const`
   -> `const BookStoreManager& Application::bookStoreManager() const`
* added `BookStoreManager& Application::bookStoreManager()`
* changed ProcessorApi call for changes at Store from
   `… ProcessorApi::Book::…(const Processor*, …)`
   -> `… ProcessorApi::Book::…(Processor*,, …)`
   to access `Application& Processor::app()`

JBenda committed:
add Processor test for MarlinMT

JBenda committed:
Fixed wrong flag value, and operator

JBenda committed:
fixed fixed type with template

(collision detection for BookStore)

JBenda committed:
changed range in TestHistogram for more hits

JBenda committed:
Added <store> node, for MarlinBook configurations

* renamed StoreOutputFile to OutputFile and moved in <store> Node

JBenda committed:
added more options, rename node to bookstore

**parameter** for <bookstore>
* OutputFile  file to store root file from BookStore, if not set creates
  file MarlinMT_[pid].root in the temporary directory of the system.
* DefaultMemoryLayout set default layout to store objects
   - Copy : use MultiCopy
   - Share : use MultiShared
   - Default : auto detect: MultiShared
* StoreByDefault determine should a new object be stored
   - YES : yes
   - NO : no
   - DEFAULT : yes, may changes later

JBenda committed:
added bitwise not operator to Flags for Masks

JBenda committed:
Now used Single memory layout if concurrency is 1

JBenda committed:
move initialization of concurrency above BookStore

JBenda commented:
Should be fine now :), added some configuration, Release notes should be up to date.

JBenda commented:
found miss labeled branch, no single mode support at the moment, will be fixed soon

JBenda committed:
Fixed wrong flag and wrong variable usage

JBenda commented:
fixed, should be fine, I will start implementing some kind of test 
for the next PR than :)

rete commented on cmake/MarlinMacros.cmake:
```diff
@@ -246,7 +247,13 @@ function( MARLIN_ADD_PROCESSOR_TEST test_name )
   # because the MARLIN_DLL contains generator expressions
-  file( READ ${PROJECT_SOURCE_DIR}/cmake/runmarlin.cmake.in TEST_FILE_CONTENT )
+	set (RUN_INPUT_FILE "${PROJECT_SOURCE_DIR}/cmake")
+	if (ARG_PARALLEL)
```
It would be better to pass the cmake target in script configuration instead of duplicating a cmake script for a change of a few characters.

JBenda replied:
sure, introduce PROCESSOR_TEST_TARGET

rete commented on source/framework/include/marlin/BookStoreManager.h:
```diff
@@ -0,0 +1,128 @@
+  class BookStoreManager {
+    struct ParameterNames {
+      static constexpr char OutPutFile[] = "OutputFile" ;
+      static constexpr char DefaultMemoryLayout[] = "DefaultMemoryLayeut" ;
```
Typo !!!

rete commented on source/framework/include/marlin/BookStoreManager.h:
```diff
@@ -0,0 +1,128 @@
+    /// path to file to store objects
+    std::filesystem::path                _storeFile{""};
+    /// default flag, used if flag == BookFlags::Default. 
+    /// Default is shared, store. Change is steering file with: store::DefaultMemoryLayeut and store::StoreByDefault.
```
Typo

rete commented on source/framework/src/BookStoreManager.cc:
```diff
@@ -0,0 +1,238 @@
+          _storeFile = "";
+        }
+        std::stringstream strs{};
+        strs << "MarlinMT_" << std::this_thread::get_id() << ".root";
```
You should get the PID instead of the thread id here.

rete commented on source/framework/src/BookStoreManager.cc:
```diff
@@ -0,0 +1,238 @@
+    _logger = _application->createLogger( "BookStoreManager" ) ;
+
+    std::shared_ptr<StringParameters> paras =  app->bookStoreParameters();
+    if (!paras) {
```
You also need a default value for the file name here

JBenda replied:
right :sweat_smile:  added belowe

rete commented on source/framework/src/BookStoreManager.cc:
```diff
@@ -0,0 +1,238 @@
+      _defaultFlag = BookFlags::MultiShared | BookFlags::Store ;  
+    } else {
+      _storeFile =
+        paras->getValue<std::string>(ParameterNames::OutPutFile, "");
```
OutPut -> Output

rete commented on source/framework/src/BookStoreManager.cc:
```diff
@@ -0,0 +1,238 @@
+        strs << "MarlinMT_" << std::this_thread::get_id() << ".root";
+        _storeFile /= strs.str(); 
+        _logger->log<WARNING>()
+          << "no output file for store defined, will write in temporary file: "
```
It is not a temporary file, just the name will be generated

JBenda replied:
no MarlinMT_[pid].root in execution directory

rete commented on test/steer/histogram.xml:
```diff
@@ -1,14 +1,21 @@
+  <bookstore>
+    <parameter name="OutputFile">histTest.root</parameter>
+    <parameter name="DefaultMemoryLayout">Copy</parameter>
+    <parameter name="StoreByDefault">YES</parameter>
```
Use a boolean here, `true` or `false`. The default would be if you don't specify anything

JBenda replied:
Flag will be removed.

JBenda committed:
removed runmarlinmt

added MARLIN_BIN_TARGET to set Marlin or MarlinMT

JBenda committed:
fixed type in interface

JBenda committed:
fixed internal  typos

JBenda committed:
removed StoreByDefault Option

default behavior is to store booked objects.

JBenda committed:
changed default output file behavior

default output file now is in base directory and used the pid for unic
numbering

JBenda committed:
No output file when no object is stored

JBenda committed:
rename output file for TestHistogram proper

rete commented:
Anything else to add ?

JBenda commented:
I don't think so
